tools/src/main/java/org/apache/kafka/tools/ManifestWorkspace.java

```diff
@@ -0,0 +1,573 @@
+        Objects.requireNonNull(pluginLocation, "pluginLocation must be non-null");
+        Objects.requireNonNull(manifestState, "manifestState must be non-null");
+        if (!Files.exists(pluginLocation)) {
+            throw new TerseException(pluginLocation + " does not exist");
+        }
+        if (!Files.isWritable(pluginLocation)) {
+            throw new TerseException(pluginLocation + " is not writable");
+        }
+        for (Map.Entry<PluginType, Set<String>> manifest : manifestState.entrySet()) {
+            PluginType type = manifest.getKey();
+            Set<String> elements = manifest.getValue();
+            Path manifestPath = pluginLocation.resolve(MANIFEST_PREFIX).resolve(type.superClass().getName());
+            rewriteManifestFile(dryRun, manifestPath, elements);
+        }
+    }
+
+    /**
+     * Rewrite a file, creating it and any parent directories if it does not exist, deleting it if specified
+     *
+     * @param dryRun              True if the rewrite should be applied, false if it should be simulated.
+     * @param filePath            Path to file which should be rewritten.
+     * @param elements            Class names of plugins which should appear in the manifest
+     */
+    private void rewriteManifestFile(boolean dryRun, Path filePath, Set<String> elements) throws IOException, TerseException {
+        Objects.requireNonNull(filePath, "filePath must be non-null");
+        Objects.requireNonNull(elements, "elements must be non-null");
+        Path writableFile = getWritablePath(dryRun, filePath);
+        if (elements.isEmpty()) {
+            if (Files.exists(filePath)) {
+                out.printf("\tDelete\t%s%n", filePath);
+                if (!dryRun) {
+                    Files.delete(writableFile);
+                }
+            }
+            // TODO: delete parent directories if empty up to the containing directory
+        } else {
+            Path parent = filePath.getParent();
+            if (parent != null) {
+                if (Files.exists(parent)) {
+                    if (!Files.isWritable(parent)) {
+                        throw new TerseException(parent + " is not writable");
+                    }
+                } else {
+                    out.printf("\tCreate\t%s%n", parent);
+                    if (!dryRun) {
+                        Files.createDirectories(parent);
+                    }
+                }
+            }
+            if (!Files.exists(filePath)) {
+                out.printf("\tCreate\t%s%n", filePath);
+            }
+            out.printf("\tWrite\t%s with content %s%n", filePath, elements);
+            if (!dryRun) {
+                try (OutputStream stream = new BufferedOutputStream(Files.newOutputStream(
+                        writableFile,
+                        StandardOpenOption.CREATE,
+                        StandardOpenOption.WRITE,
+                        StandardOpenOption.TRUNCATE_EXISTING
```

Comment:
Just for my edification, can you confirm my understanding of this logic?

We unconditionally wipe any existing manifest files at this point, even if they contain entries for plugins that were either present but not loadable, or simply not present.

This is safe because we still search for all manifest entries earlier on while doing reflection and service loader scanning, and optionally hold onto all manifest entries (even for unloadable/non-present plugins) during that time, before including them in the contents of the manifest file we write here.

Is this correct?

Reply:
Yes everything you mentioned is correct.

Reply:
Great, thanks 👍